File: LeetCode/0774.Minimize-Max-Distance-To-Gas-Station/Minimize-Max-Distance-To-Gas-Station.py
class Solution(object):
    def minmaxGasDist(self, stations, K):
        """
        :type stations: List[int]
        :type K: int
        :rtype: float
        """
        left, right = 0, stations[-1] - stations[0]
        while left + 10e-6 < right:
            mid = (left + right) / 2.0
            if self.check(stations, mid) > K:
                left = mid
            else:
                right = mid 
        return right
    
    def check(self, stations, d):
        res = 0
        for i in range(0, len(stations) - 1):
            res += math.ceil((stations[i + 1] - stations[i]) / d) - 1
        #print(d, res)
        return res
        

File: LeetCode/0774.Minimize-Max-Distance-To-Gas-Station/Minimize-Max-Distance-To-Gas-Station.java

class Solution{
    public double minmaxGasDist(int[] A, int k) {
        double L = 0, R = 100000000;
        while (R - L > 0.000001) {
            double mid = L + (R - L) / 2;
            if (tooLarge(A, mid, k)) L = mid;
            else R = mid;
        }
        return R; // we got k > target k
    }

    public boolean tooLarge(int[] A, double dist, int targetK) {
        int cnt = 0;
        for (int i = 0; i < A.length - 1; i++) {
            cnt += (A[i + 1] - A[i]) / dist;
        }
        return cnt > targetK;
    }
}

